### PR TITLE
[IMP] (website_)sale,*: readonly routes and methods

### PR DESCRIPTION
**Pull request by Feyensv**
Mark readonly routes/methods so that those
requests can be handled by readonly database replicas.

task-4167831

See also:
odoo/enterprise#74022

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website_sale/controllers/delivery.py [hunk 1]
```diff
@@ -128,7 +135,13 @@ def website_sale_set_pickup_location(self, pickup_location_data):
         order_sudo = request.website.sale_get_order()
         order_sudo._set_pickup_location(pickup_location_data)
 
-    @route('/website_sale/get_pickup_locations', type='jsonrpc', auth='public', website=True)
+    @route(
+        '/website_sale/get_pickup_locations',
+        type='jsonrpc',
+        auth='public',
+        website=True,
+        readonly=True,
```
Comment (Feyensv) [inline addons/website_sale/controllers/delivery.py:143]: NB: currently not readonly when click & collect is installed & running.
Will be fixed by #187608

**Review by ltinel — APPROVED** [review]
Thanks! Did you check the logs to make sure that these methods are indeed read-only? :)

**Comment by Feyensv** [comment]
> Thanks! Did you check the logs to make sure that these methods are indeed read-only? :)

The logs? I think those flags are mostly to say whether the route/method was readonly or not, not a feedback whether there were no write or not.
I mostly checked the method based on my understanding of the code and logic, to label places where there is no reason to write anything in the database.

File: addons/sale/controllers/product_configurator.py [hunk 2]
```diff
@@ -124,6 +124,7 @@ def sale_product_configurator_create_product(self, product_template_id, ptav_ids
         type='jsonrpc',
         auth='user',
         methods=['POST'],
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/sale/controllers/product_configurator.py:126]: What do POST/GET have to do with readonly? :thinking:

[the remark this answers is not in the document]
Comment (ltinel) [reply addons/sale/controllers/product_configurator.py:126]: Write requests should be restricted to POST (see e.g. [1]), but there's no need to restrict read requests. In Odoo, I believe RPCs are configured to send POST requests by default, so restricting read requests to POST doesn't cause any issues, but it's not necessarily a good idea IMO.

[1] https://docs.google.com/presentation/d/1IGlq6KPRfsd3EQUKVja-pX9DaCWe8qavB-JmZBILnWM/edit#slide=id.g255d973e400_0_738

**Comment by ltinel** [comment]
> > Thanks! Did you check the logs to make sure that these methods are indeed read-only? :)
> 
> The logs? I think those flags are mostly to say whether the route/method was readonly or not, not a feedback whether there were no write or not. (Also, I only noticed this information recently, not sure it was present when I made the PR) I mostly checked the method based on my understanding of the code and logic, to label places where there is no reason to write anything in the database.

I had a look at the Discord post and it's actually worse to flag a write request as read-only (instead of leaving it unflagged) since the request will be made twice in that case. And apparently you can check the logs for such occurrences. But maybe they made a check for such logs in runbot? (and if they didn't, they probably should :sweat_smile: )

**Comment by Feyensv** [comment]
Dropped two readonly on `website_sale/controllers/delivery.py` as they rely on `sale_get_order` which can sometimes trigger updates on the cart.

@robodoo r+

**Pull request closed without merging**